### DISCUSSION TITLE
Robot goes in hardware fault when using VOCAB_IM_COMPLIANT motorControlImpedance

### DISCUSSION THREAD

**Discussion by caicaixia** [robotology/community#329] [Q&A] [answered]
Dear Sir,

Today I tried the tutorial demo motorControlImpedance for the impedance control of iCub.

It is a very simple demo:
```c++
if (times%2)
        {
             // set the elbow joint only in compliant mode
             ictrl->setPositionMode(3);
             iint->setInteractionMode(3,VOCAB_IM_COMPLIANT);
             // set new reference positions
             command[0]=-50;
             command[1]=20;
             command[2]=-10;
             command[3]=60;
        }
        else
        {
             // set the elbow joint in stiff mode
            ictrl->setPositionMode(3);
             iint->setInteractionMode(3,VOCAB_IM_STIFF);
             // set new reference positions
             command[0]=-20;
             command[1]=40;
             command[2]=-10;
             command[3]=30;  
        }
```

However, I have a few problems:
1: After icub went to the `VOCAB_IM_COMPLIANT`, then the elbow joint will show a hardware fault. Then even it changes back to `VOCAB_IM_STIFF` mode, its seems like joint3 is not able to move again.

![compliant_error](https://user-images.githubusercontent.com/18044886/50821167-7b249680-1369-11e9-984e-c849ead5d251.png)

2: What is the mean of new reference positions? Joint3 will go to the new reference position or not? From the demo, I saw the command[3]=60; or command[3]=30; are never reached, it seems like the elbow joint is not controlled. This is maybe due to problem 1.

So how can I go back to position control stiff mode after I did compliant mode?  I mean how let the icub assume the normal position control?

**Comment by julijenv** [comment]
hi @caicaixia ,
could please tell me which robot your are using?
could you try to move the ebow joint from the 2 extreme values with the sliding bar in order to see if it is due to mechanical issues or software/control issues?!?
let us know

**Comment by caicaixia** [comment]
Hello @julijenv 

1:  I am working in iCubSingapore01;
2:  I tried to move the elbow joint from the 2 extreme values with the sliding bar, its no problem;
3:  I also tested other joints for  the `VOCAB_IM_COMPLIANT` mode:
     `iint->setInteractionMode(1,VOCAB_IM_COMPLIANT);`

I set joint 1,  then joint 0,1,2  all become compliant, i guess this is because those 3 joints are belong to shoulder togehter.   This test has the same hardware fault problem after the `VOCAB_IM_COMPLIANT` mode.
![comliant_error2](https://user-images.githubusercontent.com/18044886/50823987-541d9300-1370-11e9-88b4-fb9c1e17e285.png)

**Comment by julijenv** [comment]
Hello @caicaixia ,
I'm not the expert of the software part but I believe that it is the way you use the function 

> motorControlImpedance

 that lead to this problem, I will let my colleague @pattacini to get an eye to that. But I fear you may need to insert the entire code of the while loop inside a reply to get more inside of the problem you may face and giving you the right tip to use the function.
Please consider copy paste your entire code for the use of the function to help the resolution.
You may alternatively send the file you use.
thx in advance

**Comment by caicaixia** [comment]
Hello @julijenv ,

Thanks for ur reply.  However the `motorControlImpedance` is a tutorial demo, its not my code.
I also tried to modified some place to check the solution, but havn't figure out.  That's why I come here to open the issue.

Check the deomo code here:
https://github.com/robotology/icub-tutorials/blob/master/src/motorControlImpedance/tutorial_arm_joint_impedance.cpp

http://www.icub.org/doc/icub-main/tutorial__arm__joint__impedance_8cpp_source.html


Thanks very much.

**Comment by julijenv** [comment]
Hi @caicaixia ,
my colleague will return to you ASAP, I guess the thread will be moved to another repository cox it is not related to any hardware issue.
Just as a matter of clarity, are you using the overall demo file or just a part of it?!?!? ( I mean https://github.com/robotology/icub-tutorials/blob/master/src/motorControlImpedance/tutorial_arm_joint_impedance.cpp)

**Comment by pattacini** [comment]
Hi @caicaixia 

As per our contributing guidelines of [icub-support][1], I moved your issue here in [QA][2].
We are going to triage it soon, stay tuned.

(please, in the future avoid assigning issues directly to specific people since triaging is a maintainers' task)

[1]: https://github.com/robotology/icub-support/blob/master/.github/SUPPORT.md
[2]: https://github.com/robotology/QA/issues

**Comment by caicaixia** [comment]
Hi @julijenv,  I just downloaded this demo and use the whole demo.

**Comment by caicaixia** [comment]
Hi @pattacini  Sorry for ur trouble.  I WILL remember in future.

**Comment by pattacini** [comment]
@caicaixia it might be that you did not launch `wholeBodyDynamics` (`WBD` in short) prior to running the tutorial. `WBD` is required to deal with the F/T computations. For safety reasons, if `WBD` is unavailable, the joints will turn red within the GUI signaling the problem. To get them back functional, you have first to put them in `idle` (yellow color) and then select `run`.

You can find more information on this visiting http://wiki.icub.org/wiki/Force_Control, although the page is not really up-to-date with our most recent software. For this, @randaz81 might chime in giving further hints and heads-up.

**Comment by caicaixia** [comment]
Hello @pattacini , 

1:   I tried the run `wholeBodyDynamics`, but it has the same error. The join went in hardware fault.
I run the nodes as:
```xml
 <name>Tutorial on Impedance Control iCub</name>
  <dependencies>
      <port>/icub/yarprobotinterface</port>
  </dependencies>
  <module>
      <name>yarplogger</name>
      <parameters>--start --no_stop</parameters>
      <node>icubsrv</node>
  </module>
  <module>
	<name>wholeBodyDynamics</name>
	<parameters>--no_legs</parameters>
	<node>icubsrv</node>
	<stdio>icubsrv</stdio>
	<tag>wholeBodyDynamicsTag</tag>
  </module>
  <module>
	<name>gravityCompensator</name>
	<parameters>--no_legs</parameters>
	<node>icubsrv</node>
	<stdio>icubsrv</stdio>
	<tag>gravityCompensatorTag</tag>
  </module>
  <module>
      <name>motorControlImpedance</name>
      <parameters>--robot icub</parameters>
      <environment>YARP_FORWARD_LOG_ENABLE=1</environment>
      <node>icubsrv</node>
  </module>
  <connection>
        <from>/icub/right_arm/analog:o</from>
        <to>/wholeBodyDynamics/right_arm/FT:i</to>
        <protocol>udp</protocol>
  </connection>
```


Here is my yarpmanger:
![compliant_3](https://user-images.githubusercontent.com/18044886/50876803-cf7f5300-1409-11e9-9300-641a81063e97.png)


2:  In code,  I also tried to send joints in IDLE/RUN (the same as clicking the button), but it is also not working :(   I don't know what can I do anymore.
```c++
// set the elbow joint in stiff mode
//             ictrl->setControlMode(3, VOCAB_CM_FORCE_IDLE); // click the Idle button
             ictrl->setControlMode(3, VOCAB_CM_POSITION);
             iint->setInteractionMode(3,VOCAB_IM_STIFF);
```

3:  My purpose is to do a human compliant interaction with iCub,  after the contact, the icub can go back to his work again (like go to the desired position).
Similar to this video:  https://www.youtube.com/watch?v=RhtVOyKuL_0
https://www.youtube.com/watch?v=LMGSok5tN4A

**Comment by traversaro** [comment, marked as answer]
Hi @caicaixia, I suspect you are missing a few connections between the robot and the `wholeBodyDynamics` module. I suggest you to remove manual connections involving any port with the `/wholeBodyDynamics` prefix and just pass the `--autoconnect` option to both `wholeBodyDynamics` and `gravityCompensator` module. 
In particular, I suggest to just use the following application template: https://github.com/robotology/icub-main/blob/4f5881acafab693f763dd277f87a8996fdea01a3/app/iCubStartup/scripts/iCubStartup.xml.template#L68 , eventually disabling `iKinGazeCtrl` and `iKinCartesianSolver` if you are not using them. 
If you have further problems, please also report the output of the `wholeBodyDynamics` and `gravityCompensator` modules that you should see in the `yarplogger`, thanks!

**Comment by caicaixia** [comment]
hello @traversaro @pattacini,  
Thanks for the help.  The demo seems working now.  It doesn't go to the HARDWARE_FAULT anymore after the compliant mode :)

**Comment by pattacini** [comment]
Glad to hear that!
